server/src/internalClusterTest/java/org/elasticsearch/get/ShardMultiGetFomTranslogActionIT.java

```diff
@@ -0,0 +1,149 @@
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Objects;
+
+import static org.elasticsearch.action.get.ShardMultiGetFromTranslogUtil.getFailures;
+import static org.elasticsearch.action.get.ShardMultiGetFromTranslogUtil.getLocations;
+import static org.elasticsearch.action.get.ShardMultiGetFromTranslogUtil.getResponses;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class ShardMultiGetFomTranslogActionIT extends ESIntegTestCase {
+    public void testShardMultiGetFromTranslog() throws Exception {
+        assertAcked(
+            prepareCreate("test").setSettings(
+                Settings.builder()
+                    .put("index.refresh_interval", -1)
+                    // A ShardMultiGetFromTranslogAction runs only Stateless where there is only one active indexing shard.
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            ).addAlias(new Alias("alias").writeIndex(randomFrom(true, false, null)))
+        );
+        ensureGreen();
+
+        var response = getFromTranslog(indexOrAlias(), "1", "2", "3");
+        var multiGetShardResponse = response.multiGetShardResponse();
+        assertThat(getLocations(multiGetShardResponse).size(), equalTo(3));
+        assertThat(getFailures(multiGetShardResponse).size(), equalTo(3));
+        assertTrue(getFailures(multiGetShardResponse).stream().allMatch(Objects::isNull));
+        assertThat(getResponses(multiGetShardResponse).size(), equalTo(3));
+        // There hasn't been any switches from unsafe to safe map
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+
+        var indexResponse = client().prepareIndex("test")
+            .setId("1")
+            .setSource("field1", "value1")
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.NONE)
+            .get();
+        response = getFromTranslog(indexOrAlias(), "1", "2", "3");
+        multiGetShardResponse = response.multiGetShardResponse();
+        assertThat(getLocations(multiGetShardResponse).size(), equalTo(3));
+        assertThat(getFailures(multiGetShardResponse).size(), equalTo(3));
+        assertTrue(getFailures(multiGetShardResponse).stream().allMatch(Objects::isNull));
+        var getResponses = getResponses(multiGetShardResponse);
+        assertThat(getResponses.size(), equalTo(3));
+        assertNotNull(getResponses.get(0));
+        assertThat(getResponses.get(0).getId(), equalTo("1"));
+        assertThat(getResponses.get(0).getVersion(), equalTo(indexResponse.getVersion()));
```

Comment:
I think it might be nice to also assert the returned "field1" value == "value1" here and in all corresponding points below, especially when changed to "value2".